### ISSUE TITLE
Adding a client and user property programmatically via SQL init file fails

### ISSUE THREAD

**Issue by RNCTX** [babelouest/glewlwyd#154]
**Describe the issue**
Hello again! Another issue...

I'm trying to write an Ansible script to automatically deploy Glewlwyd with an extra field added to users and clients.  I've reverse engineered the process following the docs and diff'ing the database between the tables without the extra field and the tables with the extra field, and then written an Ansible script to recreate the state with the extra field by modifying the postgres init SQL file.

However this does not work, the webapp hangs at a "loading" animation after this database and config migration. I'm wondering if perhaps there is some sort of CRC check on the webapp files that I'm not aware of? Perhaps for the first run?

Doing all of this manually works, but doing it via Ansible does not.

**To Reproduce**
```
    - name: Set Glewlwyd admin password, set roles fields in initial databse SQL script...
      vars:
        admin_pass: "{{ lookup('password', '~/ansible-admin-password.txt length=32 chars=ascii_letters,digits') }}"
        roles_client: "('database', 0, 'database', 'Database backend', '{\"use-glewlwyd-connection\":true,\"data-format\":{\"redirect_uri\":{\"multiple\":true,\"read\":true,\"write\":true},\"authorization_type\":{\"multiple\":true,\"read\":true,\"write\":true},\"client_secret\":{\"read\":true,\"write\":true},\"pubkey\":{\"read\":true,\"write\":true},\"jwks\":{\"convert\":\"jwks\",\"read\":true,\"write\":true},\"jwks_uri\":{\"read\":true,\"write\":true},\"post_logout_redirect_uris\":{\"multiple\":true,\"read\":true,\"write\":true},\"token_endpoint_auth_method\":{\"multiple\":false,\"read\":true,\"write\":true},\"roles\":{\"multiple\":false,\"read\":true,\"write\":true}}}', 0);"
        roles_user: "('database', 0, 'database', 'Database backend', '{\"use-glewlwyd-connection\":true,\"data-format\":{\"picture\":{\"multiple\":false,\"read\":true,\"write\":true,\"profile-read\":true,\"profile-write\":true},\"roles\":{\"multiple\":false,\"read\":true,\"write\":true,\"profile-read\":false,\"profile-write\":false}}}', 0);"
      replace:
        path: /usr/share/glewlwyd/docs/database/init.postgre.sql
        after: "{{ item.after }}"
        regexp: "{{ item.regexp }}"
        replace: "{{ item.replace }}"
        backup: "{{ true if index == 0 else false }}"
      loop:
        - { after: "CREATE TABLE gs_oauth2_session", regexp: "^(INSERT INTO g_client_module_instance.*?VALUES\\s).*?$", replace: "\\1{{ roles_client }}" }
        - { after: "CREATE TABLE gs_oauth2_session", regexp: "^(INSERT INTO g_user_module_instance.*?VALUES\\s).*?$", replace: "\\1{{ roles_user }}" }
        - { after: "INSERT INTO g_client_module_instance", regexp: "^(INSERT INTO g_user \\(gu_username, gu_name, gu_password, gu_email, gu_enabled\\) VALUES \\('admin', 'The Administrator', crypt\\(').*?(', gen_salt\\('bf'\\)\\), '', 1\\);)$", replace: "\\1{{ admin_pass }}\\2" }
      loop_control:
        index_var: index

    - name: Migrate initial Glewlwyd database...
      vars:
        authdb_name: "authdb"
        authdb_user: "authdbuser"
        authdb_pass: "{{ lookup('password', '~/ansible-authdb_user-database-password.txt length=32 chars=ascii_letters,digits') }}"
      postgresql_query:
        db: "{{ authdb_name }}"
        login_user: "{{ authdb_user }}"
        login_password: "{{ authdb_pass }}"
        path_to_script: /usr/share/glewlwyd/docs/database/init.postgre.sql

    - name: Copy Glewlwyd webapp configs...
      template:
        src: "{{ item.src }}"
        dest: "{{ item.dest }}"
        mode: 0644
        backup: True
      loop:
        - { src: "includes/glewlwyd/templates/config_json.j2", dest: "/usr/share/glewlwyd/webapp/config.json" }
        - { src: "includes/glewlwyd/templates/translations_en.j2", dest: "/usr/share/glewlwyd/webapp/locales/en/translations.json" }
        - { src: "includes/glewlwyd/templates/translations_fr.j2", dest: "/usr/share/glewlwyd/webapp/locales/fr/translations.json" }
        - { src: "includes/glewlwyd/templates/translations_nl.j2", dest: "/usr/share/glewlwyd/webapp/locales/nl/translations.json" }
```
**Expected behavior**
Glewlwyd runs with the modified user and client field present

**Screenshots**

![image](https://user-images.githubusercontent.com/13775820/87346221-f050c900-c516-11ea-8438-0401226d3755.png)

**System (please complete the following information):**
 - OS/Environment **Ubuntu 20.04**
 - Browser used **Chrome 83, I tested Safari as well**
 - Glewlwyd Version **2.3.1**
 - Source installation **deb packages**
 - If applicable, what option did you use to build Glewlwyd **n/a**

**Additional context**

A diff on database dumps with Ansible installing the field vs manually installing the field:

```
$ diff ansible_installed_extra_field.sql manually_installed_extra_field.sql
2599c2599
< 1	admin	The Administrator		$redactredactredactredact	1
---
> 1	admin	The Administrator		$redactredactredactredact	1
2652a2653
> 1	{SHA512}redactredactredactredactredactredactredactrdactredactredactredactredact==	Mozilla/5.0 (Linux; ...	2020-07-13 18:34:00+00	1	1
2660a2662
> 1	1	\N	2020-08-12 18:34:00+00	2020-07-13 18:34:00+00	0	1
3078c3080
< SELECT pg_catalog.setval('public.g_user_session_gus_id_seq', 1, false);
---
> SELECT pg_catalog.setval('public.g_user_session_gus_id_seq', 1, true);
3085c3087
< SELECT pg_catalog.setval('public.g_user_session_scheme_guss_id_seq', 1, false);
---
> SELECT pg_catalog.setval('public.g_user_session_scheme_guss_id_seq', 1, true);
```

A diff on the postgre init files, the top modified by Ansible and the bottom the original one :

```
$ diff init.postgre.sql init.postgre.sql.22*

640,642c640,642
< INSERT INTO g_user_module_instance (gumi_module, gumi_order, gumi_name, gumi_display_name, gumi_parameters, gumi_readonly) VALUES ('database', 0, 'database', 'Database backend', '{"use-glewlwyd-connection":true,"data-format":{"picture":{"multiple":false,"read":true,"write":true,"profile-read":true,"profile-write":true},"roles":{"multiple":false,"read":true,"write":true,"profile-read":false,"profile-write":false}}}', 0);
< INSERT INTO g_client_module_instance (gcmi_module, gcmi_order, gcmi_name, gcmi_display_name, gcmi_parameters, gcmi_readonly) VALUES ('database', 0, 'database', 'Database backend', '{"use-glewlwyd-connection":true,"data-format":{"redirect_uri":{"multiple":true,"read":true,"write":true},"authorization_type":{"multiple":true,"read":true,"write":true},"client_secret":{"read":true,"write":true},"pubkey":{"read":true,"write":true},"jwks":{"convert":"jwks","read":true,"write":true},"jwks_uri":{"read":true,"write":true},"post_logout_redirect_uris":{"multiple":true,"read":true,"write":true},"token_endpoint_auth_method":{"multiple":false,"read":true,"write":true},"roles":{"multiple":false,"read":true,"write":true}}}', 0);
< INSERT INTO g_user (gu_username, gu_name, gu_password, gu_email, gu_enabled) VALUES ('admin', 'The Administrator', crypt('redactedredactedredacted', gen_salt('bf')), '', 1);
---
> INSERT INTO g_user_module_instance (gumi_module, gumi_order, gumi_name, gumi_display_name, gumi_parameters, gumi_readonly) VALUES ('database', 0, 'database', 'Database backend', '{"use-glewlwyd-connection":true,"data-format":{"picture":{"multiple":false,"read":true,"write":true,"profile-read":true,"profile-write":true}}}', 0);
> INSERT INTO g_client_module_instance (gcmi_module, gcmi_order, gcmi_name, gcmi_display_name, gcmi_parameters, gcmi_readonly) VALUES ('database', 0, 'database', 'Database backend', '{"use-glewlwyd-connection":true,"data-format":{"redirect_uri":{"multiple":true,"read":true,"write":true},"authorization_type":{"multiple":true,"read":true,"write":true},"client_secret":{"read":true,"write":true},"pubkey":{"read":true,"write":true},"jwks":{"convert":"jwks","read":true,"write":true},"jwks_uri":{"read":true,"write":true},"post_logout_redirect_uris":{"multiple":true,"read":true,"write":true},"token_endpoint_auth_method":{"multiple":false,"read":true,"write":true}}}', 0);
> INSERT INTO g_user (gu_username, gu_name, gu_password, gu_email, gu_enabled) VALUES ('admin', 'The Administrator', crypt('password', gen_salt('bf')), '', 1);
```

A diff on the translations en file, the top modified by Ansible and the bottom the original one:

```
diff translations.json translations.json.231*
164,165d163
<     "user-roles": "User's role claim",
<     "user-roles-ph": "ex: web_user or web_admin",
233,234d230
<     "client-roles": "A user role claim",
<     "client-roles-ph": "ex: web_user or web_admin",
1086c1082
< }
\ No newline at end of file
---
> }
```

A diff on the config.json file, the top modified by Ansible and the bottom the original one:

```
diff config.json config.json.230*
66,72d65
<       "name": "roles",
<       "profile": false,
<       "edit": true,
<       "label": "admin.user-roles",
<       "placeholder": "admin.user-roles-ph",
<       "forceShow": true,
<     },{
196,202d188
<     },{
<       "name": "roles",
<       "profile": false,
<       "edit": true,
<       "label": "admin.client-roles",
<       "placeholder": "admin.client-roles-ph",
<       "forceShow": true,
```

**Comment by babelouest** [comment]
I believe your issue lies in a tiny but very sneaky character you added by mistake in one of the updated files.

TL;DR: you added an extra comma in the config.json in the `forceShow` property.

i.e., replace `"forceShow": true,` with `"forceShow": true` in both `roles` properties in your updated config.json file.

I was gonna start a long, boring and pedantic explanation about how JSON isn't fault-tolerant, but I would hate myself after that ;-)

**Comment by RNCTX** [comment]
You are correct, I just found it and was coming to type the answer, and I also found an extra json string that gets added on the backend that differs from the db init script:

`"sector_identifier_uri":{"multiple":false,"read":true,"write":true}`

After adding those it works, thanks again ;)

**Comment by babelouest** [comment]
And to answer a side question, there is no CRC verification on Glewlwyd's files at runtime, I don't it would be useful, and it could lower the performance.

I could sign the packages and tarballs though, this would be cheaper and as efficient...
